While composing an email when using Gmail in Safari, using a double space creates an extra line/visible character in space between paragraphs in email.

rr-codes opened the pull request:
#### da96b048f1a990c60856466d44440b90e7e16257
<pre>
While composing an email when using Gmail in Safari, using a double space creates an extra line/visible character in space between paragraphs in email.
<a href="https://bugs.webkit.org/show_bug.cgi?id=242651">https://bugs.webkit.org/show_bug.cgi?id=242651</a>
rdar://96150129

Reviewed by Ryosuke Niwa and Wenson Hsieh.

Prior to r294523, selection updates were performed synchronously, so there is no pending
update when layout is being performed, and Gmail&apos;s node insertion/removal does not result
in a selection update.

In Gmail. when composing a new email, adding a newline causes `setSelection` to be called,
which schedules a selection update by setting `m_pendingSelectionUpdate = true`.

Then, the `keydown` event listener is fired, and a zero-delay timer on the Gmail JS side
is invoked. A text node is then inserted into the document via JS.

A layout update is then triggered, which causes a selection update, which calls `recomputeCaretRect`.

The problem is then that `m_previousCaretNode` gets set based on the current selection,
however that selected node then gets removed from the document. Since `m_previousCaretNode`
is a removed node, no rect is returned and so invalidation never happens leading to a
stale caret still being displayed.

This PR fixes this by checking for this failure case, and then ensuring that
`m_previousCaretNode` is not in a bad state and does not represent a removed node.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::appliedEditing):
* Source/WebCore/editing/Editor.h:
(WebCore::Editor::setIsProcessingKeyCommand):
(WebCore::Editor::isProcessingKeyCommand const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::executeKeypressCommandsInternal):

Canonical link: <a href="https://commits.webkit.org/252683@main">https://commits.webkit.org/252683@main</a>
</pre>

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -1159,6 +1159,8 @@ void Editor::appliedEditing(CompositeEditCommand& command)
         OptionSet<FrameSelection::SetSelectionOption> options;
         if (command.isDictationCommand())
             options.add(FrameSelection::DictationTriggered);
+        if (isProcessingKeyCommand())
```
Could this method be defined on `CompositeEditCommand` rather than on `Editor`? Similar to `isDictationCommand()` above, I wonder if this check could be `command.isKeyCommand()`.

cc @whsieh

whsieh replied:
I think having the state on `CompositeEditCommand` would be nice, but it may be tricky, since what we really care about is "was the `Editor::Command` created with `CommandFromMenuOrKeyBinding`".

whsieh replied:
That aside, I think there's (broadly speaking) two viable strategies we can take, to fix the bug at hand:
• Make selection changes during user-triggered edit commands actually use the `UserTriggered` flag when setting the selection.
• ...or, make it so that scheduling an async selection change never results in us painting a stale selection caret.

It seems like it might be worth investigating the second bullet a bit, to figure out why the async selection change leaves behind this painting artifact. I suspect this investigation would also give us some insight as to how we can devise a test to exercise this bug.

(It might have something to do with installing zero-delay timers that race against the next rendering update, after performing an "InsertParagraph" command)

rniwa reviewed (approved):
r=me if perf tests turn out okay.

rr-codes committed:
While composing an email when using Gmail in Safari, using a double space creates an extra line/visible character in space between paragraphs in email.
https://bugs.webkit.org/show_bug.cgi?id=242651
rdar://96150129

Reviewed by Ryosuke Niwa and Wenson Hsieh.

Prior to r294523, selection updates were performed synchronously, so there is no pending
update when layout is being performed, and Gmail's node insertion/removal does not result
in a selection update.

In Gmail. when composing a new email, adding a newline causes `setSelection` to be called,
which schedules a selection update by setting `m_pendingSelectionUpdate = true`.

Then, the `keydown` event listener is fired, and a zero-delay timer on the Gmail JS side
is invoked. A text node is then inserted into the document via JS.

A layout update is then triggered, which causes a selection update, which calls `recomputeCaretRect`.

The problem is then that `m_previousCaretNode` gets set based on the current selection,
however that selected node then gets removed from the document. Since `m_previousCaretNode`
is a removed node, no rect is returned and so invalidation never happens leading to a
stale caret still being displayed.

This PR fixes this by checking for this failure case, and then ensuring that
`m_previousCaretNode` is not in a bad state and does not represent a removed node.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::appliedEditing):
* Source/WebCore/editing/Editor.h:
(WebCore::Editor::setIsProcessingKeyCommand):
(WebCore::Editor::isProcessingKeyCommand const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::executeKeypressCommandsInternal):

Canonical link: https://commits.webkit.org/252683@main
